### ISSUE TITLE
Large --exclude-from file causes tremendous slowdown

### ISSUE THREAD

**Issue by fgouget** [borgbackup/borg#1167]
I have a lot of build directories. These mix source code, which I want to backup, and the resulting binaries which take space and which I can always rebuild. Ignoring the object files is a nobrainer with "_.o" but I don't want to ignore all "_.so" files because I have software elsewhere which I do want to backup. So for tar I was building a list of the files I explicitly want to exclude.

Reusing tar's exclude list with borg --exclude-from makes it unbearably slow, even after the initial backup. Here is a way to reproduce the problem:

```
#!/bin/sh

[ -d "repo" ] || borg init -e none repo

if [ ! -d "src" ]
then
    mkdir "src"
    i=0
    while [ $i -lt 100 ]
    do
        touch "src/$i"
        i=`expr $i + 1`
    done
fi

if [ ! -f "exclude" ]
then
    (
        i=0
        while [ $i -lt 10000 ]
        do
            echo "/home/user/some/directory/$i"
            #echo "fm:/home/user/some/directory/$i"
            #echo "sh:/home/user/some/directory/$i"
            #echo "re:^/home/user/some/directory/$i\$"
            i=`expr $i + 1`
        done
    ) >exclude
fi

set -x
# This is fast
time borg create repo::std-{now} src

# This is 10 times as slow
time borg create --exclude-from exclude repo::exclude-{now} src
```

On my computer I get:

```
+ time borg create repo::std-{now} src
0.13user 0.01system 0:00.95elapsed 15%CPU (0avgtext+0avgdata 24784maxresident)k
+ time borg create --exclude-from exclude repo::exclude-{now} src
1.69user 0.02system 0:02.40elapsed 71%CPU (0avgtext+0avgdata 41336maxresident)k
```

The more entries in the exclude file the slower borg gets in a roughly linear fashion. This seems true for all pattern types. In my case it was processing around 5 filenames per second.

Fortunately borg has more sophisticated patterns for exclusion so it's possible to drastically reduce the number of entries in the exclude file. But this comes at the cost of having one exclude file for tar and another for borg.

So this may not be high priority but it would be nice to special-case matching plain filenames to make that case fast (e.g. put them in a Python set and check the current path against it).

**Comment by ThomasWaldmann** [comment]
as we already support different kinds of patterns (see http://borgbackup.readthedocs.io/en/stable/usage.html?highlight=pattern#borg-help-patterns ), maybe it could be optimized adding just another kind of "precise match" pattern.

**Comment by ThomasWaldmann** [comment]
Ideas what you could try right now:
- you could maybe easily transform your list of N precise match filenames into 1 (or few) regex patterns automatically: A|B|C|D
- you could use the path prefix pattern style, it is less expensive than the others (but still O(N))

**Comment by ThomasWaldmann** [comment]
As an alternative idea, maybe we could one day feed the list of filenames that you actually want to backup into borg. If you can automatically generate that, that would be an alternate way.

**Comment by fgouget** [comment]
"re:" is a bit faster than "fm:", 1.1s instead of 1.69s.
"re:A|B|C" is faster still: 0.7s.
And "pp:' is the fastest at 0.5s

But none of that helps.

The only usable workaround is to change the way I generate the exclude file to go from ~125000 lines to 600 patterns. The backup goes from taking about 28 hours to completing in under 10 minutes!

**Comment by dragetd** [comment]
125k lines of excludes is pretty big. If you generate those from patterns - reusing those to create the explude patters for borg (and having this 600-line file) is the more sane approach than introducing another pattern-style in borg IMHO.

And it might improve slightly (maybe not in such an extreme case) when the multicore support is implemented. :)

**Comment by fgouget** [comment]
I have run into this problem again, this time when trying to exclude the files that are ignored by Git from the backup.

Of course the idea is to use a blanket exclude directive for things like *.o, *.lo, etc. But this helps exclude all binaries that are not easily excluded because they don't have extensions (note that not all projects put their binaries in .libs directories), and other files that are trivially rebuilt like makefiles (from Makefile.in, etc). It's pretty easy to achieve using a command like this to generate chunks of the exclude file:

( cd "$git_dir" && LANG= git clean -Xdn ) | sed -e '/^Would skip repository / d' -e "s%^Would remove %pp:$git_dir/%" | egrep -v '\.(a|la|lai|lo|o|Plo|Po|spec\.c)$'

But even so with a bunch of build trees the exclude list grows to around 10000 entries and borg goes from taking 13 minutes to taking 90 to 150 minutes, even using 'pp:' matches.

Really users should not have to do contortions to get acceptable performance from borg's filtering algorithm. Dealing with 10000 plain strings should be no issue at all.

**Comment by enkore** [comment]
So, while thinking about #2288, it occurred to me that obviously pattern and exclude files are a domain-specific programming language (DSL) that is compiled (usually to a regex which is then JITed) and the result interpreted by Borg. Now, what can a compiler do? Optimize! Borg could merge adjacent patterns of the same genus (include/exclude) and type (regex vs. prefix) for some types, e.g. adjacent regexes can be ORed together which would likely be faster ([especially with some engines](https://github.com/google/re2)) than checking many REs against a path.

**Comment by ThomasWaldmann** [comment]
Maybe we could offer a special pattern type for such use cases, let's call it "precise (exclude|include)".

```
pi:/here/is/a/full/path/to/include/no/matter/what.txt
pe:/here/is/a/full/path/to/exclude/no/matter/what.txt
```
As they are completely unconditional / context free and fully specified, we could just throw them into 2 sets, so processing them just takes 2 set operations.

**Comment by enkore** [comment]
Since we already have --include/--exclude, it should be just one pattern type; precise / exact (pr:/ex:?).

So I think there are two tickets to extract here and this one can be closed:

1.) Precise match "pattern"
2.) Optimize how excludes are processed. The current implementation is naive, as outlined above, and can be improved in many ways:

  - Adjacent patterns converted to REs can be merged
  - A different / faster RE engine could be used
  - For precise matches sets could be used
  - For prefix matches and precise matches tries could be used. I mention this one since I already experimented with this for BorgCube and tries showed extraordinarily well compact representation of many paths (e.g. pytries was able to map millions of files to numbers in about 5 MB of disk space, while the list of file names was 300+ MB).

**Comment by ThomasWaldmann** [comment]
@enkore yes, I agree with all those possibilities (and that they would yield best performance for big lists of such patterns), but:
- except the full path match, they are solving problems we did not have yet / nobody complained about
- for patterns that actually have variable parts, having big numbers of them is less likely (as each pattern matches multiple / many fs items)
- considering that, above sounds a bit overengineered :)

So, I started the simplest thing in PR #2334.
This approach needs to ignore pattern order and context for the PathFullPatterns, we can just document that (and I think this is not really a restriction in practice: if you make lists of full-path matches and you include or exclude them, it is unlikely that you want to have conflicting other patterns).

**Comment by fgouget** [comment]
Actually a new type of match may not be needed: for the 'fn' type of match the wild card characters are '?', '*' and '['. So if an entry does not contain any of these, which it likely does not, then verifying for a match can be done using a set. The only issue is that fn patterns must match the path from start to either the end or to a path separator. But as long as directories are treated recursively and pruned as soon as they match this should work.

**Comment by ThomasWaldmann** [comment]
@fgouget I also needed the new type because of a semantics change: these "full path" matches are processed first, independently of context (of all other patterns). So if a full path match says "have this file" or "do not have this file", the decision is taken, it won't consider any other includes/excludes. This is needed to have a simple and quick set lookup, otherwise it would get way more complicated and slower again.

**Comment by enkore** [comment]
That's mainly a problem due to interaction with --patterns. The fast path @fgouget mentioned still works unless there are interfering --patterns, so using it if there aren't any would pose no problems.

**Comment by ThomasWaldmann** [comment]
Checking for interference with all other patterns totally sounds like not fun (maybe expensive. maybe impossible). So I'ld rather keep this simple and not overengineered.

**Comment by enkore** [comment]
> so using it if there aren't **any** [--patterns, at all]

**Comment by enkore** [comment]
Optimizing compiler for patterns is a separate project that will need a better structure and representation than what's there now. Single changes sprinkling if's over the existing code will thus not be accepted.

**Comment by ThomasWaldmann** [comment]
This is fixed by using "pf:" patterns, added by #2334.
